src/DataWrangling/EN4/EN4.jl

```diff
@@ -0,0 +1,257 @@
+                    architecture = CPU(),
+                    inpainting = default_inpainting(metadata),
+                    mask = nothing,
+                    horizontal_halo = (7, 7),
+                    cache_inpainted_data = true)
+                    
+    field = empty_EN4_field(metadata; architecture, horizontal_halo)
+    inpainted_path = inpainted_metadata_path(metadata)
+
+    if !isnothing(inpainting) && isfile(inpainted_path)
+        file = jldopen(inpainted_path, "r")
+        maxiter = file["inpainting_maxiter"]
+
+        # read data if generated with the same inpainting
+        if maxiter == inpainting.maxiter
+            data = file["data"]
+            close(file)
+            copyto!(parent(field), data)
+            return field
+        end
+
+        close(file)
+    end
+
+    download_dataset(metadata)
+    path = metadata_path(metadata)
+    ds = Dataset(path)
+    shortname = short_name(metadata)
+
+    if variable_is_three_dimensional(metadata)
+        data = ds[shortname][:, :, :, 1]
+        data = reverse(data, dims=3)
+    else
+        data = ds[shortname][:, :, 1]
+    end        
+
+    close(ds)
+    
+    # Convert data from Union(FT, missing} to FT
+    FT = eltype(field)
+    data[ismissing.(data)] .= 1e10 # Artificially large number!
+    data = if location(field)[2] == Face # ?
+        new_data = zeros(FT, size(field))
+        new_data[:, 1:end-1, :] .= data
+        new_data    
+    else
+        data = Array{FT}(data)
+    end
+    
+    set!(field, data)
+    fill_halo_regions!(field)
+
+    if !isnothing(inpainting)
+        # Respect user-supplied mask, but otherwise build default EN4 mask.
+        if isnothing(mask)
+            mask = EN4_mask(metadata, architecture; data_field=field)
+        end
+
+        # Make sure all values are extended properly
+        name = string(metadata.name)
+        date = string(metadata.dates)
+        dataset = summary(metadata.dataset)
+        @info string("Inpainting ", dataset, " ", name, " data from ", date, "...")
+        start_time = time_ns()
+        
+        inpaint_mask!(field, mask; inpainting)
+        fill_halo_regions!(field)
+
+        elapsed = 1e-9 * (time_ns() - start_time)
+        @info string(" ... (", prettytime(elapsed), ")")
+    
+        # We cache the inpainted data to avoid recomputing it
+        @root if cache_inpainted_data
+            file = jldopen(inpainted_path, "w+")
+            file["data"] = on_architecture(CPU(), parent(field))
+            file["inpainting_maxiter"] = inpainting.maxiter
+            close(file)
+        end
+    end
+
+    return field
+end
+
+# Fallback
+EN4_field(var_name::Symbol; kw...) = EN4_field(EN4Metadata(var_name); kw...)
+
+function inpainted_metadata_filename(metadata::EN4Metadata, var::AbstractString)
+    original_filename = metadata_filename(metadata)
+    without_extension = original_filename[1:end-3]
+    return without_extension * "_" * var *"_inpainted.jld2"
```

Comment:
FYI - I added `var` as an addition to the unpainted file names, ensuring that if one file contains multiple variables, each unpainted variable is output as a separate file

Reply:
I think the `var` is included in the `metadata_filename`. Is this not the case for `EN4`?

Reply:
Yes, it is, I updated this function again so it doesn't require the extra `var` argument and can pull it directly from `metadata.name`